### PR TITLE
Faster, reduced allocation ImageProcessorImageUrlGenerator using zString

### PR DESCRIPTION
**Pull request by nzdev**
### Prerequisites

- [ ] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes 

### Description
What?
Introduce [zString](https://github.com/Cysharp/ZString), a zero allocation stringbuilder and reduce cpu time & memory allocation for building imageprocessor urls.

Why?
It's a good example of how zString can be used in place of stringbuilder to optimize code without significantly affecting readability of the code. It could be used in other parts of the codebase as well where performance matters.

How to test?
Existing unit tests pass. Benchmark test vs existing code added.

Benchmark results:

```
// * Summary *

        //        BenchmarkDotNet=v0.11.3, OS=Windows 10.0.19041
        //Intel Core i5-6300U CPU 2.40GHz(Skylake), 1 CPU, 4 logical and 2 physical cores
        // [Host]     : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 32bit LegacyJIT-v4.8.4220.0
        //  Job-FTOBUC : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 32bit LegacyJIT-v4.8.4220.0

        //InvocationCount=1  UnrollFactor=1

        //            Method |     Mean |     Error |    StdDev |   Median | Ratio | RatioSD | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
        //------------------ |---------:|----------:|----------:|---------:|------:|--------:|------------:|------------:|------------:|--------------------:|
        //  UseStringBuilder | 2.867 ms | 0.2656 ms | 0.7830 ms | 2.784 ms |  1.00 |    0.00 |           - |           - |           - |              952 KB |
        // UseZStringBuilder | 1.743 ms | 0.1417 ms | 0.3904 ms | 1.623 ms |  0.66 |    0.23 |           - |           - |           - |              360 KB |

```

### CODE REVIEW

**Comment by Shazwazza** [comment]
very cool. I wonder if this could be used in DefaultShortStringHelper.CleanCodeString. That doesn't use StringBuilder and instead uses char arrays but I wonder what performs and allocates better. There's a lot of places where StringBuilder is used though and some in reasonably critical code.

**Comment by nzdev** [comment]
@Shazwazza Possibly. Using Span<char> you can save a few bytes for DefaultShortStringHelper at the expense of  <2 us. https://github.com/nzdev/Umbraco-CMS/commit/042f7825cb09f1d7cda91461d51ec0f718c2e674

File: src/Umbraco.Web/Umbraco.Web.csproj [hunk 1]
```diff
@@ -1285,7 +1288,7 @@
     </PropertyGroup>
     <ItemGroup>
       <!-- we want to exclude all facade references ?! -->
-      <FixedReferencePath Include="@(ReferencePath)" Condition="'%(ReferencePath.FileName)' != 'System.ValueTuple' and '%(ReferencePath.FileName)' != 'System.Net.Http'" />
+      <FixedReferencePath Include="@(ReferencePath)" Condition="'%(ReferencePath.FileName)' != 'System.ValueTuple' and '%(ReferencePath.FileName)' != 'System.Net.Http'  and '%(ReferencePath.FileName)' != 'Microsoft.Bcl.AsyncInterfaces' and '%(ReferencePath.FileName)' != 'System.Buffers' and '%(ReferencePath.FileName)' != 'System.Numerics.Vectors' and '%(ReferencePath.FileName)' != 'System.Runtime.CompilerServices.Unsafe'" />
```
Comment (Shazwazza) [inline src/Umbraco.Web/Umbraco.Web.csproj:1291]: haha, this thing, like what the heck!

**Review by Shazwazza — APPROVED** [review]
I'm all for this, looks great and could replace a lot of usages of SB in other places. In critical code this can save a lot of mem/GC. For merging to netcore we will just need to be smart about where we take on dependencies but that's fine since this implementation can exist in the Infrastructure project there. Hard to see the actual code changes since it looks like methods are moved around but I don't think the actual logic has changed. If tests pass I'm good with it!

But I guess our nuspecs will need to be updated for the additional dependency. 

@nul800sebastiaan Just pinging you here about this one

File: src/Umbraco.Web/Umbraco.Web.csproj (hunk 1 above)
Comment (nzdev) [reply src/Umbraco.Web/Umbraco.Web.csproj:1291]: It's so painful. Seems like its related to WCF or something.

**Pull request closed without merging**